common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            tokens: size.get(),
+            size,
+            rate,
+            last_refresh_time: Instant::now(),
+            open: false,
+        }
+    }
+
+    /// A fully open rate limiter, to allow for ignoring rate limiting for tests
+    pub fn open() -> Self {
+        let max = NonZeroUsize::new(std::usize::MAX).unwrap();
+        Self {
+            tokens: std::usize::MAX,
+            size: max,
+            rate: max,
+            last_refresh_time: Instant::now(),
+            open: true,
+        }
+    }
+
+    /// Refill tokens based on how many seconds have passed since last refresh
+    pub(crate) fn refill(&mut self) {
+        let num_intervals = self.last_refresh_time.elapsed().as_secs();
+        if num_intervals > 0 {
+            self.add_tokens((num_intervals as usize).saturating_mul(self.rate.get()));
+
+            // We have to base everything off the original time, or we'll have drift where we slowly slow the bucket refill rate
+            self.last_refresh_time += Duration::from_secs(num_intervals);
+        }
+    }
+
+    /// Determine if an entire batch can be passed through
+    /// This is important for message based rate limiting, where the whole message has
+    /// to make it through, or else it must be rejected.  A result of `None` means it cannot
+    /// ever be allowed through, as it's bigger than the size of the bucket.
+    pub fn acquire_all_tokens(&mut self, requested: usize) -> Result<(), Option<Instant>> {
+        // Skip over if we purposely have an open throttle
+        if self.open || requested == 0 {
+            return Ok(());
+        }
+
+        // Refill if needed
+        self.refill();
+
+        if self.tokens >= requested {
+            self.deduct_tokens(requested);
+            Ok(())
+        } else {
+            Err(self.time_of_tokens_needed(requested))
+        }
+    }
+
+    /// Returns `usize` of tokens allowed.  May be less than requested.
+    /// For best effort, caller should return unused tokens with `add_tokens`
+    pub fn acquire_tokens(&mut self, requested: usize) -> Result<usize, Instant> {
+        // Skip over if we purposely have an open throttle
+        if self.open || requested == 0 {
+            return Ok(requested);
+        }
+
+        // Refill if needed
+        self.refill();
+
+        let allowed = self.deduct_tokens(requested);
+        if allowed > 0 {
+            Ok(allowed)
+        } else {
+            Err(self.time_of_next_refill())
+        }
+    }
+
+    /// Retrieve the maximum amount of tokens up to `count`
+    /// Tells us how much of the requested size we can send
+    fn deduct_tokens(&mut self, requested: usize) -> usize {
+        let tokens = self.tokens;
+        let tokens_left = tokens.saturating_sub(requested);
+        let tokens_allowed = min(tokens, requested);
+        self.tokens = tokens_left;
```

Comment:
why are we setting tokens_left and self.tokens, feels like this should just be set once and erase line 232... specifically
```
let tokens_allowed = min(self.tokens, requested)
self.tokens = self.tokens.saturating_sub(tokens_allowed)
```

Reply:
Lack of sleep :P